### PR TITLE
Benchmarks for SslStream

### PR DESCRIPTION
**Pull request by rzikm**
This PR adds benchmarks for SslStreams. Supported scenarios are:

- read-write - measures throughput of a single SslStream connection
- handshakes - measures the number of handshakes per second.

### CODE REVIEW

File: src/BenchmarksApps/SslStream/Client/Program.cs
Comment (CarnaViire) [inline src/BenchmarksApps/SslStream/Client/Program.cs:112]: I don't think it's a good idea to use OCE to cancel the last operation 🤔 You are measuring op/s, so I believe it should count the last one as well. Otherwise by not waiting for the last op to finish, you are superficially lowering the result -- because you are adding (part of the) time, but not the bytes. I think it doesn't matter that much that the benchmark itself will run a little bit longer that the asked e.g. 15s, it's better to get more accurate results.

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (CarnaViire) [inline src/BenchmarksApps/SslStream/Server/Program.cs:34]: 👀

File: scenarios/sslstream.benchmarks.yml
Comment (CarnaViire) [inline scenarios/sslstream.benchmarks.yml:86]:
```suggestion
    readyStateText: Listening on # app should write this line to output
```
We should wait until the server starts listening, before starting the client. `readyStateText` is [a signal for that](https://github.com/dotnet/crank/blob/main/src/Microsoft.Crank.Controller/benchmarks.schema.json#L102). `isConsoleApp` should be `false` (default value), because "console app", as opposed to a "server", is [considered "ready" right away](https://github.com/dotnet/crank/blob/main/src/Microsoft.Crank.Controller/benchmarks.schema.json#L154), without waiting for a signal.

File: scenarios/sslstream.benchmarks.yml
Comment (CarnaViire) [inline scenarios/sslstream.benchmarks.yml:12]: What happens if it's "None"? Does it refuse to connect, or is it not encrypted? Is this multi-value?

File: src/BenchmarksApps/SslStream/Client/ClientOptions.cs
Comment (CarnaViire) [inline src/BenchmarksApps/SslStream/Client/ClientOptions.cs:30]: I think I had to use the explicit binder in HttpClient benchmarks, because I had to manually parse the headers parameter from `string[]` into the list of (name, value) pairs... I think it should be possible to use an automatic binder in your case.

File: src/BenchmarksApps/SslStream/Client/Program.cs
Comment (CarnaViire) [inline src/BenchmarksApps/SslStream/Client/Program.cs:167]:
```suggestion
        Task.WaitAll(writeTask, readTask);
```
Maybe?

File: scenarios/sslstream.benchmarks.yml
Comment (CarnaViire) [inline scenarios/sslstream.benchmarks.yml:13]: Is "certificate source" a term that is commonly used? 😅 because this option was confusing me, for quite some time I thought it was about where the _benchmark app_ should get the cert 😅

File: scenarios/sslstream.benchmarks.yml
Comment (CarnaViire) [inline scenarios/sslstream.benchmarks.yml:80]: Let's maybe have the code here that is in "ready for merge" state, just so we won't forget and accidentally merge the file with your local folders? 😄

File: scenarios/sslstream.benchmarks.yml
Comment (CarnaViire) [reply scenarios/sslstream.benchmarks.yml:12]: (should default be "Default"? 😄)

File: scenarios/sslstream.benchmarks.yml
Comment (rzikm) [reply scenarios/sslstream.benchmarks.yml:13]: We use something similar in S.N.S unit tests

https://github.com/dotnet/runtime/blob/77d643d8678d0cae1144557f106bb693a1cbca25/src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamMutualAuthenticationTest.cs#L36-L41

I am open for suggestions.

File: src/BenchmarksApps/SslStream/Client/ClientOptions.cs
Comment (rzikm) [reply src/BenchmarksApps/SslStream/Client/ClientOptions.cs:30]: as part of the binding, the code loads up the X509Certificate (path + optional password), so that the instance is already part of the resulting ClientOptions.

File: scenarios/sslstream.benchmarks.yml
Comment (rzikm) [reply scenarios/sslstream.benchmarks.yml:12]: The naming is a bit unfortunate and has historical baggage, but Default is actually `Ssl3 | Tls` and is deprecated. I will see if I can remove it from the list.

https://github.com/dotnet/runtime/blob/77d643d8678d0cae1144557f106bb693a1cbca25/src/libraries/System.Net.Primitives/src/System/Net/SecureProtocols/SslEnumTypes.cs#L10-L26

Comment (rzikm) [reply scenarios/sslstream.benchmarks.yml:12]: the None (= 0) value lets OS decide (i.e. highest supported TLS version)

Comment (CarnaViire) [reply scenarios/sslstream.benchmarks.yml:12]: We don't have to use the actual SslProtocols enum here. It might be beneficial to limit the values to the ones we/endusers actually use. I.e. use a new enum that is later mapped to SslProtocols (esp given that you have an explicit binder anyway). And I'd suggest then not to use the confusing naming like "None" and "Default" 😅 So only have a small, like, 2-3 value enum here.

(btw on a side note, i don't think that letting OS decide stuff is a good thing from benchmarking perspective, it's not reproducible enough 😅 it's better to set everything explicitly 🤔)

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (rzikm) [reply src/BenchmarksApps/SslStream/Server/Program.cs:34]: Not sure what you are trying to say :D

File: scenarios/sslstream.benchmarks.yml
Comment (CarnaViire) [reply scenarios/sslstream.benchmarks.yml:13]: I was thinking about it, and maybe something like "CertSelection"/"CertSelectionType"?

and this is only for the server now? you don't pass it to the client, but I see there's an option to do that in the future. then it might make sense to separate the variables by name, because the selection can be different between client and server.

```suggestion
  ServerCertSelection: "CertContext" # <Callback|Property|CertContext>
  ClientCertSelection: "CertContext" # <Callback|Collection|CertContext>
```

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (CarnaViire) [inline src/BenchmarksApps/SslStream/Server/Program.cs:126]: I think it will be better to count and report the number of errors (maybe even separately per kind). I believe, currently we've made the client to always finish "gracefully" in normal circumstances 🤔 am I missing something?

Also, even if we decide to ignore the last socket exception -- it should be the _last_ exception _only_ -- exceptions in the middle are wrong and should be reported to metrics.

_Just logging_ is not enough -- when the benchmark completes successfully (i.e. there were no unhandled exceptions), you normally don't see server's (and client's) logs, only the metrics. (esp. given that the end-goal is to add this as a power bi graph)

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (CarnaViire) [inline src/BenchmarksApps/SslStream/Server/Program.cs:117]: Expanding the comment on the socket exception -- it made me think about unhandled exceptions -- I think we should treat some errors as "unrecoverable", meaning the benchmark itself must be set up incorrectly for us to get such an error. So we need to log, but should not prevent the exception from bubbling up. Then the benchmark will be aborted, and the error together with the logs will be reported in crank output.

E.g. if we suddenly negotiated an unknown protocol, it looks like something went wrong and we should not proceed further :)

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (CarnaViire) [inline src/BenchmarksApps/SslStream/Server/Program.cs:32]: Re: 👀:

It's just I didn't understand why we need this at all, so it looked like something leftover from debugging. Now I think I see the idea behind it. I'm still not fully convinced this needs to be this way (and I'm not sure crank uses the same event to kill the server?)

I'm not 100% sure, but I believe, the way the crank will work is to wait until the client exits, and only then kill the server... So, in my view, the only active operation at this point should be listening on a socket, and I think it shouldn't need a complicated termination, so just crank killing the process will do the trick? But I might be wrong, so LMK if I'm missing something

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (CarnaViire) [inline src/BenchmarksApps/SslStream/Server/Program.cs:86]: should we start reading at all, if `options.ReceiveBufferSize` is 0? Maybe we need to do the same check as with the writing task, and then wait on both of the tasks?

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (CarnaViire) [inline src/BenchmarksApps/SslStream/Server/Program.cs:109]: Do I understand correctly that disposing the SslStream will also close the socket?

File: src/BenchmarksApps/SslStream/Client/Program.cs
Comment (CarnaViire) [inline src/BenchmarksApps/SslStream/Client/Program.cs:150]: The way the two of them are not set together atomically (and also, most possibly, during some operation that currently is running) worries me a bit 😅

in HttpClient benchmarks, it's explicitly done [between iterations](https://github.com/aspnet/Benchmarks/blob/444b7bf3b253f85ef44e1efb89824fbdc5426352/src/BenchmarksApps/HttpClientBenchmarks/Clients/HttpClient/Program.cs#L303-L314), so I wonder whether we might want to do something like that here?

**Comment by CarnaViire** [comment]
cc @liveans -- FYI, we are discussing here some ideas and concerns regarding a benchmark implementation (e.g. running and finishing a scenario, exception handling, etc), some of the comments might be useful

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (rzikm) [reply src/BenchmarksApps/SslStream/Server/Program.cs:109]: Yes, the SslStream is set to dispose the inner NetworkStream which in turn is set to own the Socket. see `EstablishSslStreamAsync`

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (rzikm) [reply src/BenchmarksApps/SslStream/Server/Program.cs:32]: It probably is not necessary at this point anymore, it was useful during debugging. On the other hand, I don't think it hurts or complicates the code much.

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (rzikm) [reply src/BenchmarksApps/SslStream/Server/Program.cs:117]: We can't easilly bubble it up out of Main because we hand off incomming connections to a task on threadpool, but I can add a metric for errors.

File: src/BenchmarksApps/SslStream/Client/Program.cs
Comment (wfurt) [inline src/BenchmarksApps/SslStream/Client/Program.cs:203]: I'm not sure if we need all this but I'm fine if we keep it. I would expect the differences would be minimal. I would expect the biggest difference would be certificate or not.

File: scenarios/sslstream.benchmarks.yml
Comment (wfurt) [inline scenarios/sslstream.benchmarks.yml:12]: should ve leave it to None as default? It may make it less predictable but it would track default use nicely IMHO

File: src/BenchmarksApps/SslStream/Client/Program.cs
Comment (wfurt) [inline src/BenchmarksApps/SslStream/Client/Program.cs:150]: what is typical duration? We should make sure the read/write phase is much longer than handshake

Comment (wfurt) [reply src/BenchmarksApps/SslStream/Client/Program.cs:150]: or would that happen in warmup phase?

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (wfurt) [reply src/BenchmarksApps/SslStream/Server/Program.cs:36]: I don't know what is precedent in the code base. I personally would be fine with leaving debug hooks but they should probably deserve comments or documentation .... so new persons can find and use them

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (wfurt) [inline src/BenchmarksApps/SslStream/Server/Program.cs:123]: any reason why not to use existing values? There are some optimizations for common values (like h2) and this is going to be somewhat more expensive. I'm not sure if that matters but I feel it would be nice if the benchmark represents common scenario.

File: scenarios/sslstream.benchmarks.yml
Comment (rzikm) [reply scenarios/sslstream.benchmarks.yml:12]: It was default before, but @carnaviire argued that we should be explicit about the version for consistency.

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (rzikm) [reply src/BenchmarksApps/SslStream/Server/Program.cs:36]: After we spoke offline about reporting connection errors, I think we need to leave this in to have opportunity to actually report them.

File: src/BenchmarksApps/SslStream/Client/Program.cs
Comment (rzikm) [reply src/BenchmarksApps/SslStream/Client/Program.cs:150]: I copied default values from HttpClient benchmarks, both warmup and duration are 15s by default.

Comment (rzikm) [reply src/BenchmarksApps/SslStream/Client/Program.cs:150]: handshake is not accounted for in either of those in the readwrite scenario

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (rzikm) [reply src/BenchmarksApps/SslStream/Server/Program.cs:123]: I wanted to have the server agnostic about what scenario (i.e. only client chooses the scenario and you don't have to restart the server) and didn't want to do the selection by sending an explicit message from the client.

We potentially could assign "h2" and "h3" values to "handshake" and "readwrite" scenarios. to make sure the optimization kicks in

File: src/BenchmarksApps/SslStream/Client/Program.cs
Comment (rzikm) [reply src/BenchmarksApps/SslStream/Client/Program.cs:206]: I would also expect differences in Context vs non-context. And once we implement TLS resume for client certs (https://github.com/dotnet/runtime/issues/94561), then I would expect cert callback to be the slowest.

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (rzikm) [reply src/BenchmarksApps/SslStream/Server/Program.cs]: Changed to report number of errors.

File: src/BenchmarksApps/SslStream/Client/Program.cs
Comment (rzikm) [reply src/BenchmarksApps/SslStream/Client/Program.cs:198]: I'd prefer to keep this in as a knob to allow us benchmarking TLS resume vs non-resume scenarios, e.g. when we implement it on linux with client certs.

**Comment by rzikm** [comment]
@CarnaViire, @wfurt  PTAL

File: scenarios/sslstream.benchmarks.yml
Comment (CarnaViire) [inline scenarios/sslstream.benchmarks.yml:95]: hmmm I think the first one should be "host" instead of "secondaryAddress" 🤔 so that if you change the address, you change it in variables only

```suggestion
    arguments: "--host {{host}} --port {{serverPort}} --receive-buffer-size {{receiveBufferSize}} --send-buffer-size {{sendBufferSize}} --tls-version {{tlsVersion}} --cert-selection {{clientCertSelection}} --scenario {{scenario}} --warmup 00:00:{{warmup}} --duration 00:00:{{duration}}"

```

File: src/BenchmarksApps/SslStream/Client/Program.cs
Comment (CarnaViire) [reply src/BenchmarksApps/SslStream/Client/Program.cs:177]: This is good point; we might want to add a comment just so we would remember the reasoning in the future 🤔 you don't have to do this in this PR though

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (CarnaViire) [inline src/BenchmarksApps/SslStream/Server/Program.cs:57]: Does this _actually_ report the error count when there are any? 🤔 I mean, within the Crank environment

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (CarnaViire) [inline src/BenchmarksApps/SslStream/Server/Program.cs:145]: this needs to be an interlocked operation

File: src/BenchmarksApps/SslStream/Server/Program.cs
Comment (rzikm) [reply src/BenchmarksApps/SslStream/Server/Program.cs:57]: Yes

File: src/BenchmarksApps/SslStream/Client/Program.cs
Comment (CarnaViire) [inline src/BenchmarksApps/SslStream/Client/Program.cs:89]: Hmmm this wasn't _exactly_ what I had in mind, I remember that the idea behind the HttpClient benchmark approach was to not actually _stop_ between warmup and the "real" metric collection -- so that any additional cost of creating/growing a state machine etc will not affect the metrics (that being said, I don't know _exactly_ how much it affects the metrics; we had this more as a theoretical discussion with Geoff and Cory when I did the initial work on HttpClient benchmarks)

We can leave it as it is right now, I don't want to hold this back much longer, but we might want to address this later.

Comment (CarnaViire) [reply src/BenchmarksApps/SslStream/Client/Program.cs:89]: (now I see that read-write scenarios seem to follow the pattern from HttpClient benchmarks, but handshakes don't 😅)

File: scenarios/sslstream.benchmarks.yml
Comment (CarnaViire) [inline scenarios/sslstream.benchmarks.yml:46]: I think this and `client-write` scenarios should also include `host: "{{secondaryAddress}}"`, shouldn't they? 🤔

**Comment by CarnaViire** [comment]
Ping @rzikm? 😊

File: src/BenchmarksApps/SslStream/Client/Program.cs
Comment (rzikm) [reply src/BenchmarksApps/SslStream/Client/Program.cs:82]: For measuring the handshake duration, it does not make sense IMO to run multiple clients, since we are essentially measuring latency

Comment (rzikm) [reply src/BenchmarksApps/SslStream/Client/Program.cs:82]: we can expand this to multiple threads in the future if we find use for it

File: src/BenchmarksApps/SslStream/Client/Program.cs
Comment (rzikm) [reply src/BenchmarksApps/SslStream/Client/Program.cs]: Converted this to use similar approach as in httpclient benchmarks, and even made it report percentiles

File: scenarios/sslstream.benchmarks.yml
Comment (rzikm) [reply scenarios/sslstream.benchmarks.yml]: fixed

File: src/BenchmarksApps/SslStream/Client/Program.cs
Comment (rzikm) [reply src/BenchmarksApps/SslStream/Client/Program.cs:185]: added comment

**Comment by CarnaViire** [comment]
LGTM, but why there is a submodule added? 😅
image

**Comment by rzikm** [comment]
No Idea how it got there, should be gone now

**Review by CarnaViire — APPROVED** [review]

**Pull request merged**
